plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/controllers/VagrantLaunchUtility.java

```diff
@@ -120,32 +119,93 @@ private ILaunchConfigurationWorkingCopy findLaunchConfig(IServer s, String launc
 		return ret.toArray(new String[ret.size()]);
 	}
 
-	public static String[] call(String rootCommand, String[] args, File vagrantDir,
-			Map<String, String> env) throws IOException {
+	public static String[] call(String rootCommand, String[] args, File vagrantDir, Map<String, String> env)
+			throws IOException, TimeoutException {
+		return call(rootCommand, args, vagrantDir, env, 30000);
+	}
+
+	public static String[] call(String rootCommand, String[] args, File vagrantDir, Map<String, String> env,
+			int timeout) throws IOException, TimeoutException {
+
 		String[] envp = (env == null ? null : convertEnvironment(env));
 
-		List<String> result = new ArrayList<>();
 		List<String> cmd = new ArrayList<>();
 		cmd.add(rootCommand);
 		cmd.addAll(Arrays.asList(args));
-		Process p = Runtime.getRuntime().exec(cmd.toArray(new String[0]),
-				envp, vagrantDir);
-		BufferedReader buff = new BufferedReader(
-				new InputStreamReader(p.getInputStream()));
-		try {
-			if (p.waitFor() == 0) {
-				String line;
-				while ((line = buff.readLine()) != null) {
-					result.add(line);
-				}
-			} else {
-				return new String[0];
+		final Process p = Runtime.getRuntime().exec(cmd.toArray(new String[0]), envp, vagrantDir);
+
+		InputStream errStream = p.getErrorStream();
+		InputStream inStream = p.getInputStream();
+
+		Integer exitCode = ThreadUtils.runWithTimeout(timeout, new Callable<Integer>() {
+			@Override
+		 	public Integer call() throws Exception {
+				return p.waitFor();
 			}
-		} catch(InterruptedException ie) {
-			// ignore
+		});
+		
+		try ( StreamGobbler inGobbler = new StreamGobbler(inStream); 
+				StreamGobbler errGobbler = new StreamGobbler(errStream)) {
+			
+			new Thread(inGobbler).start();
+			new Thread(errGobbler).start();
+
+			if( exitCode == null ) {
+				// Timeout reached
+				inGobbler.cancel();
+				errGobbler.cancel();
+				throw new TimeoutException(getTimeoutError(inGobbler.getLines(), errGobbler.getLines()));
+			}
+			List<String> ret = inGobbler.getLines();
+			return (String[]) ret.toArray(new String[ret.size()]);
 		}
-		return result.toArray(new String[0]);
 	}
 	
+	private static String getTimeoutError(List<String> output, List<String> err) {
+		StringBuilder msg = new StringBuilder();
+		msg.append("Process output:\n");
+		output.forEach(line -> msg.append("   ").append(line));
+		err.forEach(line -> msg.append("   ").append(line));
+		return msg.toString();
+	}
 	
+	private static class StreamGobbler implements Runnable, AutoCloseable {
+		private Scanner inScanner;
+		private CopyOnWriteArrayList<String> lines;
+		private boolean canceled = false;
+		private BufferedInputStream is;
+		public StreamGobbler(InputStream inStream) {
+			lines = new CopyOnWriteArrayList<>();
```

Comment:
CopyOnWriteArrayList makes complete copy of its array when adding each element. If the number of lines may be more than 100, it will impact performance.
If getLines() be started with cancel() before return lines, then no item will be added to the list after getLines() is called.